src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicStream.cs

```diff
@@ -778,31 +830,80 @@ private static uint HandleEvent(State state, ref StreamEvent evt)
 
         private static unsafe uint HandleEventRecv(State state, ref StreamEvent evt)
         {
-            StreamEventDataReceive receiveEvent = evt.Data.Receive;
-            for (int i = 0; i < receiveEvent.BufferCount; i++)
+            ref StreamEventDataReceive receiveEvent = ref evt.Data.Receive;
+
+            if (receiveEvent.BufferCount == 0)
             {
-                state.ReceiveQuicBuffers.Add(receiveEvent.Buffers[i]);
+                // This is a 0-length receive that happens once reads are finished (via abort or otherwise).
+                // State changes for this are handled elsewhere.
+                return MsQuicStatusCodes.Success;
             }
 
+            int readLength;
+
             bool shouldComplete = false;
             lock (state)
             {
-                if (state.ReadState == ReadState.None)
+                switch (state.ReadState)
                 {
-                    shouldComplete = true;
-                }
-                if (state.ReadState != ReadState.ConnectionClosed)
-                {
-                    state.ReadState = ReadState.IndividualReadComplete;
+                    case ReadState.None:
+                        // ReadAsync() hasn't been called yet. Stash the buffer so the next ReadAsync call completes synchronously.
+
+                        if ((uint)state.ReceiveQuicBuffers.Length < receiveEvent.BufferCount)
+                        {
+                            QuicBuffer[] oldReceiveBuffers = state.ReceiveQuicBuffers;
+                            state.ReceiveQuicBuffers = ArrayPool<QuicBuffer>.Shared.Rent((int)receiveEvent.BufferCount);
+
+                            if (oldReceiveBuffers.Length != 0) // don't return Array.Empty.
```

Comment:
Should we assert somehow the old buffers were fully consumed?

Reply:
There's no need to do that. Unconsumed data arrives again from the point we've stopped. New RECV event wouldn't come until we call `EnableReceive()`, and we call it only after we've consumed as much as we could and said so to msquic in `ReceiveComplete(taken)`, so new event will have all the remaining data.

Reply:
Maybe a comment wouldn't hurt?